File: sdks/typescript/src/webhooks/waas.ts
import { assertRuntimeContract } from "../core/contract.js";
import { CregisClientError } from "../core/errors.js";
import { waasWebhooks } from "../generated/waas/webhooks.js";
import type {
  AddressDepositCallbackNotification,
  PayoutCallbackNotification,
  PayoutExternalVerificationCallbackNotification,
  WithdrawalCallbackNotification,
} from "./types.js";
import { verifyProjectWebhook } from "./validation.js";

export class CregisWaasCallbackHandler {
  public static readonly CALLBACK_SUCCESS = "success";
  public static readonly EXTERNAL_VERIFICATION_APPROVE = "ok";
  public static readonly EXTERNAL_VERIFICATION_DENY = "deny";
  readonly #apiKey: string;

  public constructor(apiKey: string) {
    if (typeof apiKey !== "string" || apiKey.trim() === "") {
      throw new CregisClientError("API Key is required");
    }
    this.#apiKey = apiKey;
  }

  public handleDepositCallback(rawBody: string): AddressDepositCallbackNotification {
    return this.#verify<AddressDepositCallbackNotification>(
      rawBody,
      waasWebhooks.depositCallback,
      "WaaS deposit callback",
    );
  }

  public handlePayoutCallback(rawBody: string): PayoutCallbackNotification {
    return this.#verify<PayoutCallbackNotification>(
      rawBody,
      waasWebhooks.payoutCallback,
      "WaaS payout callback",
    );
  }

  public handlePayoutExternalVerificationCallback(
    rawBody: string,
  ): PayoutExternalVerificationCallbackNotification {
    return this.#verify<PayoutExternalVerificationCallbackNotification>(
      rawBody,
      waasWebhooks.payoutExternalVerificationCallback,
      "WaaS payout external verification callback",
    );
  }

  public handleWithdrawalCallback(rawBody: string): WithdrawalCallbackNotification {
    return this.#verify<WithdrawalCallbackNotification>(
      rawBody,
      waasWebhooks.withdrawalCallback,
      "WaaS withdrawal callback",
    );
  }

  #verify<T>(
    rawBody: string,
    contract: (typeof waasWebhooks)[keyof typeof waasWebhooks],
    name: string,
  ): T {
    const payload = verifyProjectWebhook(rawBody, this.#apiKey);
    assertRuntimeContract(payload, contract.schema, contract.schemas, name);
    return payload as T;
  }
}

File: sdks/python/src/cregis/webhooks/waas.py
"""WaaS callback verification and contract parsing."""

from __future__ import annotations

from typing import Type, TypeVar

from pydantic import BaseModel, ValidationError

from cregis.errors import CregisClientError, CregisContractError
from cregis.generated.waas.models import (
    AddressDepositCallbackNotification,
    PayoutCallbackNotification,
    PayoutExternalVerificationCallbackNotification,
    WithdrawalCallbackNotification,
)
from cregis.webhooks.validation import RawBody, verify_project_webhook

WebhookModel = TypeVar("WebhookModel", bound=BaseModel)


class CregisWaasCallbackHandler:
    CALLBACK_SUCCESS = "success"
    EXTERNAL_VERIFICATION_APPROVE = "ok"
    EXTERNAL_VERIFICATION_DENY = "deny"

    def __init__(self, api_key: str) -> None:
        if not isinstance(api_key, str) or not api_key.strip():
            raise CregisClientError("API Key is required")
        self._api_key = api_key

    def handle_deposit_callback(self, raw_body: RawBody) -> AddressDepositCallbackNotification:
        return self._verify(raw_body, AddressDepositCallbackNotification, "WaaS deposit callback")

    def handle_payout_callback(self, raw_body: RawBody) -> PayoutCallbackNotification:
        return self._verify(raw_body, PayoutCallbackNotification, "WaaS payout callback")

    def handle_payout_external_verification_callback(
        self, raw_body: RawBody
    ) -> PayoutExternalVerificationCallbackNotification:
        return self._verify(
            raw_body,
            PayoutExternalVerificationCallbackNotification,
            "WaaS payout external verification callback",
        )

    def handle_withdrawal_callback(self, raw_body: RawBody) -> WithdrawalCallbackNotification:
        return self._verify(
            raw_body,
            WithdrawalCallbackNotification,
            "WaaS withdrawal callback",
        )

    def _verify(
        self,
        raw_body: RawBody,
        model: Type[WebhookModel],
        name: str,
    ) -> WebhookModel:
        payload = verify_project_webhook(raw_body, self._api_key)
        try:
            return model.model_validate(payload, strict=True)
        except (ValidationError, TypeError, ValueError) as exc:
            raise CregisContractError(name, exc) from exc
